xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,173 @@
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: dmp.SwapOp, rewriter: PatternRewriter, /):
+        # remove op if it contains no swaps
+        if op.swaps is None or len(op.swaps) == 0:
+            rewriter.erase_matched_op(False)
+            return
+        assert isinstance(op.input_stencil, OpResult)
+
+        # currently only works for 3-dimensional stencils
+        assert all(len(swap.size) == 3 for swap in op.swaps)
+
+        # dmp should decompose from (x,y,z) to (1,1,z)
+        assert all(swap.size[:2] == (1, 1) for swap in op.swaps)
+
+        # check that size is uniform
+        uniform_size = op.swaps.data[0].size[2]
+        assert all(swap.size[2] == uniform_size for swap in op.swaps)
+
+        assert isa(
+            op.input_stencil.type,
+            memref.MemRefType[Attribute] | stencil.TempType[Attribute],
+        )
+        assert isa(
+            t_type := op.input_stencil.type.get_element_type(), TensorType[Attribute]
+        )
+
+        # when translating swaps, remove third dimension
+        prefetch_op = csl_stencil.PrefetchOp(
+            input_stencil=op.input_stencil.op,
+            topo=op.topo,
+            swaps=[
+                csl_stencil.ExchangeDeclarationAttr(swap.neighbor[:2])
+                for swap in op.swaps
+            ],
+            result_type=memref.MemRefType(
+                TensorType(t_type.get_element_type(), (uniform_size,)),
+                (len(op.swaps),),
+            ),
+        )
+
+        # a little hack to get around a check that prevents replacing a no-results op with an n-results op
+        rewriter.replace_matched_op(prefetch_op, new_results=[])
+
+        # uses have to be retrieved *before* the loop because of the rewriting happening inside the loop
+        uses = list(op.input_stencil.uses)
+
+        # csl_stencil.prefetch, unlike dmp.swap, has a return value. This is added as the last arg
+        # to stencil.apply, before rebuilding the op and replacing stencil.access ops by csl_stencil.access ops
+        # that reference the prefetched buffers (note, this is only done for neighbor accesses)
+        for use in uses:
+            if not isinstance(use.operation, stencil.ApplyOp):
+                continue
+            apply_op = use.operation
+
+            # arg_idx points to the stencil.temp type whose data is prefetched in a separate buffer
+            arg_idx = apply_op.args.index(op.input_stencil)
+
+            # add the prefetched buffer as the last arg to stencil.access
+            apply_op.region.block.insert_arg(
+                prefetch_op.result.type, len(apply_op.args)
+            )
+
+            # rebuild stencil.apply op
+            r_types = [r.type for r in apply_op.results]
+            assert isa(r_types, Sequence[stencil.TempType[Attribute]])
+            new_apply_op = stencil.ApplyOp.get(
+                list(apply_op.args) + [prefetch_op.result],
+                apply_op.region.clone(),
```

Comment:
I think detatching the region would be cheaper here.

Reply:
Nice